crates/rune-macros/src/any.rs

```diff
@@ -780,4 +786,222 @@ where
+                        #static_type_mod::#ty.hash
+                    }
+
+                    #[inline]
+                    fn type_info() -> #type_info {
+                        #type_info::StaticType(#static_type_mod::#ty)
+                    }
+                }
+
+                #[automatically_derived]
+                impl #impl_generics #maybe_type_of for #ident #type_generics #where_clause {
+                    #[inline]
+                    fn maybe_type_of() -> Option<#full_type_of> {
+                        Some(<Self as #type_of>::type_of())
+                    }
+                }
+            })
+        } else {
+            None
+        };
+
+        let any_ref = attr.builtin.is_none().then(|| {
+            let type_hash = type_hash.into_inner();
+
+            let make_hash = if !generic_names.is_empty() {
+                quote!(#hash::new_with_type_parameters(#type_hash, #hash::parameters([#(<#generic_names as #type_of>::type_hash()),*])))
+            } else {
+                quote!(#hash::new(#type_hash))
+            };
+
+            Some(quote! {
+                #[automatically_derived]
+                impl #impl_generics #type_hash_trait for #ident #type_generics #where_clause {
+                    fn type_hash() -> #hash {
+                        #make_hash
+                    }
+                }
+
+                #[automatically_derived]
+                impl #impl_generics #any_ref for #ident #type_generics #where_clause {
+
+                }
+
+                #[automatically_derived]
+                impl #impl_generics #unsafe_to_value for &#ident #type_generics #where_clause {
+                    type Guard = ();
+
+                    unsafe fn unsafe_to_value(self) -> #vm_result<(#value, Self::Guard)> {
+                        let any_obj = #shared::new(#any_obj::new_projection(self));
+                        #vm_result::Ok((#value::from(any_obj), ()))
+                    }
+                }
+
+                #[automatically_derived]
+                impl #impl_generics #unsafe_to_value for &mut #ident #type_generics #where_clause {
+                    type Guard = ();
+
+                    unsafe fn unsafe_to_value(self) -> #vm_result<(#value, Self::Guard)> {
+                        let any_obj = #shared::new(#any_obj::new_projection_mut(self));
+                        #vm_result::Ok((#value::from(any_obj), ()))
+                    }
+                }
+
+                #[automatically_derived]
+                impl #impl_generics #unsafe_to_ref for #ident #type_generics #where_clause {
+                    type Guard = #raw_ref;
+
+                    unsafe fn unsafe_to_ref<'__rune_g>(value: #value) -> #vm_result<(&'__rune_g Self, Self::Guard)> {
+                        let (projection, guard) = #vm_try!(value.into_projection());
+                        let projection = projection.as_ref();
+                        #vm_result::Ok((#lifetime::get(projection), guard))
```

Comment:
This _appears_ to work, although I have not tested it thoroughly yet. By invoking the trait here, we're able to keep the `Projection<T>` type out of closure arguments, and avoid having to include it in `impl`s.